controller/rpc/scanner.go

```diff
@@ -15,90 +15,142 @@ import (
 )
 
 type scannerAct struct {
-	scanner  *share.CLUSScanner
-	scanning uint32
+	scanner            *share.CLUSScanner
+	activeScannerTasks int // Indicate how much scanner task is currently run
 }
 
-var scanners map[string]*scannerAct = make(map[string]*scannerAct)
-var scanMutex sync.RWMutex
+type ScannerManager struct {
+	activeScanners map[string]*scannerAct
+	maxConns       int
+	mutex          sync.RWMutex
+	availableCh    chan struct{}
+}
+
+var ScannerMgr *ScannerManager
+
+// defaultScannerCount ensures that the availableCh channel has sufficient capacity to handle task signals without blocking.
+// Default number of scanners to prevent availableCh channel from being overwhelmed by tasks
+var defaultScannerCount = 15
+
+func NewScannerManager(maxConns int) {
+	ScannerMgr = &ScannerManager{
+		activeScanners: make(map[string]*scannerAct),
+		maxConns:       maxConns,
+		availableCh:    make(chan struct{}, defaultScannerCount*maxConns),
+	}
+}
 
-func AddScanner(scanner *share.CLUSScanner) {
-	scanMutex.Lock()
-	defer scanMutex.Unlock()
-	scanners[scanner.ID] = &scannerAct{scanner: scanner}
+func (mgr *ScannerManager) signalAvailability() {
```

Comment:
Can you elaborate more about the purpose of `availableCh`?  I'm not sure if I fully understand, but given that `signalAvailability()` would block when `defaultScannerCount*maxConns` is reached, I'm sure I definitely missed something.